File: cpp/cpp_100-999/714.cc
// C++
// leetcode 714
// https://leetcode-cn.com/problems/best-time-to-buy-and-sell-stock-with-transaction-fee/

class Solution {
public:
    int maxProfit(vector<int>& prices, int fee) {
    	int buy = INT_MIN, sell = 0;
    	for(auto &x : prices) {
    		buy = max(buy, sell - x);
    		sell = max(sell, buy + x - fee);
    	}
    	return sell;
    }
};

File: Go/Go_100-999/714.go
// Go
// leetcode 714
// https://leetcode.cn/problems/best-time-to-buy-and-sell-stock-with-transaction-fee/

func maxProfit(prices []int, fee int) int {
	cnt := []int{-100000000, 0}
	for _, x := range prices {
		cnt[0] = max(cnt[0], cnt[1]-x)
		cnt[1] = max(cnt[1], cnt[0] + x - fee)
	}
	return cnt[1]
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}